packages/protocol/contracts/L1/libs/LibProving.sol

```diff
@@ -138,87 +89,75 @@ library LibProving {
-            fc.provenAt = uint64(
-                block.timestamp.max(blk.proposedAt + config.proofTimeTarget)
-            );
-        } else {
-            fc.provenAt = uint64(block.timestamp);
-        }
+        // Put together the input for proof and signature verification
+        uint256[9] memory inputs;
 
-        if (evidence.prover != address(0) && evidence.prover != address(1)) {
-            uint256[9] memory inputs;
-
-            inputs[0] = uint256(
-                uint160(address(resolver.resolve("signal_service", false)))
-            );
-            inputs[1] = uint256(
-                uint160(
-                    address(
-                        resolver.resolve(
-                            config.chainId,
-                            "signal_service",
-                            false
-                        )
-                    )
-                )
-            );
-            inputs[2] = uint256(
-                uint160(
-                    address(resolver.resolve(config.chainId, "taiko", false))
+        inputs[0] = uint256(
+            uint160(address(resolver.resolve("signal_service", false)))
+        );
+        inputs[1] = uint256(
+            uint160(
+                address(
+                    resolver.resolve(config.chainId, "signal_service", false)
                 )
-            );
-
-            inputs[3] = uint256(blk.metaHash);
-            inputs[4] = uint256(evidence.parentHash);
-            inputs[5] = uint256(evidence.blockHash);
-            inputs[6] = uint256(evidence.signalRoot);
-            inputs[7] = uint256(evidence.graffiti);
-            inputs[8] =
-                (uint256(uint160(evidence.prover)) << 96) |
-                (uint256(evidence.parentGasUsed) << 64) |
-                (uint256(evidence.gasUsed) << 32);
-
-            bytes32 instance;
-            assembly {
-                instance := keccak256(inputs, mul(32, 9))
-            }
+            )
+        );
+        inputs[2] = uint256(
+            uint160(address(resolver.resolve(config.chainId, "taiko", false)))
+        );
 
-            (bool verified, bytes memory ret) = resolver
-                .resolve(LibUtils.getVerifierName(evidence.verifierId), false)
-                .staticcall(bytes.concat(instance, evidence.proof));
+        inputs[3] = uint256(blk.metaHash);
+        inputs[4] = uint256(evidence.parentHash);
+        inputs[5] = uint256(evidence.blockHash);
+        inputs[6] = uint256(evidence.signalRoot);
+        inputs[7] = uint256(evidence.graffiti);
+        inputs[8] =
+            (uint256(uint160(evidence.prover)) << 96) |
+            (uint256(evidence.parentGasUsed) << 64) |
+            (uint256(evidence.gasUsed) << 32);
+
+        bytes32 instance;
+        assembly {
+            instance := keccak256(inputs, mul(32, 9))
+        }
+
+        // config.proofToggleMask can support 8 different proof types on 8 bits
+        for (uint8 index; index < 8; ) {
```

Comment:
Should we do something like this?

```solidity
uint8 mask = config.proofToggleMask;

for (uint i = 0; i < blockProofs.length; i++) {
  TypedProof memory proof = blockProofs[i];
  uint8 bitMask = 1 << (proof.proofType - 1);
  require(mask & bitMask) != 0, "type_not_enabled_or_duplicated");
  _verifyTypedProof(proof);
  mask &= ~bitMask;
}
require(mask == 0, "not_all_requierd_proof_verified");
```

Reply:
I also suggest that the cooldown period is a value returned from `_verifyTypedProof`:

```solidity
uint8 mask = config.proofToggleMask;
uint64 maxCoolDownPeriod;
for (uint i = 0; i < blockProofs.length; i++) {
  TypedProof memory proof = blockProofs[i];
  uint8 bitMask = 1 << (proof.proofType - 1);
  require(mask & bitMask) != 0, "type_not_enabled_or_duplicated");
  uint64 coolDownPeriod = _verifyTypedProof(proof);
  maxCoolDownPeriod = maxCoolDownPeriod.max(coolDownPeriod);
  mask &= ~bitMask;
}
require(mask == 0, "not_all_requierd_proof_verified");
fc.provenAt = uint64(block.timestamp +maxCoolDownPeriod );
```

Reply:
Inside, LibVerifyerSGX and LibVerifyingZKP, in the verify function, we also return a cooldown period. This allows for different cooldown period for different verifiers, then we just take the max.

Reply:
> Inside, LibVerifyerSGX and LibVerifyingZKP, in the verify function, we also return a cooldown period. This allows for different cooldown period for different verifiers, then we just take the max.

Maybe min makes more sense otherwise adding a proof type can increase the cooldown, although cooldown period length will always a bit tricky and handwavy and not sure if something that can be decided per proof type.

> Question to this cooldown period: Since this affects `proofTime = provenAt - proposedAt` and cooldown only delays verification, why should not we let the the `LibVerifying.sol` do this job and revert ?

What do you mean with revert?